go/core/tracing/tracing.go

```diff
@@ -91,35 +201,103 @@ const (
 	spanTypeAttr = attrPrefix + ":type"
 )
 
-// RunInNewSpan runs f on input in a new span with the given name.
-// The attrs map provides the span's initial attributes.
+// SpanMetadata contains metadata information for creating properly annotated spans
+type SpanMetadata struct {
+	// Name is the span name
+	Name string
+	// IsRoot indicates if this is a root span
+	IsRoot bool
+	// Type represents the kind of span (e.g., "action", "flowStep")
+	Type string
+	// Subtype provides more specific categorization (e.g., "tool", "flow", "model")
+	Subtype string
+	// TelemetryLabels are arbitrary key-value pairs set directly as span attributes
+	// These can have any name (e.g., "genkit:sessionId": "123", "custom.label": "value")
+	// This matches TypeScript's telemetryLabels concept
+	TelemetryLabels map[string]string
+	// Metadata are genkit-specific metadata that get "genkit:metadata:" prefix
+	// (e.g., "subtype": "tool" becomes "genkit:metadata:subtype": "tool")
+	Metadata map[string]string
+}
+
+// RunInNewSpan runs f on input in a new span with the provided metadata.
+// The metadata contains all span configuration including name, type, labels, etc.
 func RunInNewSpan[I, O any](
 	ctx context.Context,
-	name, spanType string,
-	isRoot bool,
+	state *State,
```

Comment:
Isn't this nil most places? Can we just always operate on the global `otel.GetTracerProvider()`?